src/wasi/descriptor.rs

```diff
@@ -0,0 +1,521 @@
+            let name_str =
+                unsafe { std::str::from_utf8_unchecked(&dir_buff[off..off + len as usize]) };
+            off += len as usize; // Move past the name
+            name_jsarray.push(Value::from_string(JString::from_str(cx.clone(), name_str)?));
+        }
+        let name_jsarray = Array::from_iter_js(&cx, name_jsarray.iter())?;
+        return Ok(Value::from_array(name_jsarray));
+    }
+
+    /// The read method
+    /// Uint8Array as the buffer the first parameter
+    /// size as the second parameter, it's optional, default is the length of the buffer
+    fn read<'js>(self: Arc<Self>, cx: Ctx<'js>, args: Rest<Value<'js>>) -> Result<Value<'js>> {
+        if args.0.len() < 1 {
+            bail!(
+                "read expects 1 parameters: the buffer and size[option], Got: {} parameters.",
+                args.len()
+            );
+        }
+        let buffer = &args.0[0];
+        let null = Value::new_null(cx.clone());
+        let mut size = &null;
+        if args.0.len() > 2 {
+            size = &args.0[1];
+        }
+
+        let mut readn: i32 = 0;
+        let readn_ptr: i32 = &mut readn as *mut i32 as i32;
+        let array = buffer
+            .as_object()
+            .ok_or_else(|| anyhow!("buffer must be a array"))?;
+        let array: &TypedArray<'_, u8> = array
+            .as_typed_array()
+            .ok_or_else(|| anyhow!("buffer must be a typed array"))?;
+        let mut array_raw = array
+            .as_raw()
+            .ok_or_else(|| anyhow!("buffer get raw ptr error."))?;
+        let size = size.as_int();
+        let size = if let Some(size) = size {
+            if size > array_raw.len as i32 {
+                array_raw.len as u32
+            } else {
+                size as u32
+            }
+        } else {
+            array_raw.len as u32
+        };
+        let ioslice = vec![Iovec {
+            buf: unsafe { array_raw.ptr.as_mut() as *mut u8 as i32 },
+            buf_len: size,
+        }];
+        let rs = unsafe {
+            preview_1::fd_read(
+                self.0,
+                ioslice.as_ptr() as i32,
+                ioslice.len() as i32,
+                readn_ptr,
+            )
+        };
+        process_error(cx.clone(), rs)?;
+        if rs != 0 {
+            readn = -rs;
+        }
+        Ok(Value::new_int(cx, readn))
+    }
+
+    /// The write method
+    /// Uint8Array as the buffer the first parameter
+    /// size as the second parameter, it's optional, default is the length of the buffer
+    fn write<'js>(self: Arc<Self>, cx: Ctx<'js>, args: Rest<Value<'js>>) -> Result<Value<'js>> {
+        if args.0.len() < 1 {
+            bail!(
+                "write expects 1 parameters: the buffer and size[option], Got: {} parameters.",
+                args.len()
+            );
+        }
+        let buffer = &args.0[0];
+        let null = Value::new_null(cx.clone());
+        let mut size = &null;
+        if args.0.len() > 2 {
```

Comment:
same here? - should check for size atleast 2:
```suggestion
        if args.0.len() > 1 {
```